src/ai/backend/agent/server.py

```diff
@@ -292,6 +294,7 @@ async def update_scaling_group(self, scaling_group):
             data = tomlkit.load(f)
             data['agent']['scaling-group'] = scaling_group
         with open(cfg_src_path, 'w') as f:
+            shutil.copy(cfg_src_path, f"{cfg_src_path}.bak")
```

Comment:
It is kind-of safer to copy the file before opening it.
```suggestion
        shutil.copy(cfg_src_path, f"{cfg_src_path}.bak")
        with open(cfg_src_path, 'w') as f:
```

And let's wrap the method body using [an executor](https://docs.python.org/3.10/library/asyncio-eventloop.html#executing-code-in-thread-or-process-pools) and `asyncio.Lock` to keep it async and prevent potential races on updates.